elf_reader.go

```diff
@@ -541,28 +579,67 @@ func mapSpecFromBTF(spec *btf.Spec, name string) (*MapSpec, error) {
 			}
 
 		case "pinning":
+			if inner {
+				return nil, errors.New("inner maps can't be pinned")
+			}
+
 			pinning, err := uintFromBTF(member.Type)
 			if err != nil {
 				return nil, fmt.Errorf("can't get pinning: %w", err)
 			}
 
 			pinType = PinType(pinning)
 
-		case "key", "value":
+		case "values":
+			// The 'values' field in BTF map definitions is used for declaring map
+			// value types that are references to other BPF objects, like other maps
+			// or programs. It is always expected to be an array of pointers.
+			// TODO: Check if 'values' is the last field in the map definition.
+			valueType, err := resolveBTFArrayMacro(member.Type)
+			if err != nil {
+				return nil, fmt.Errorf("can't resolve BTF value type: %w", err)
+			}
+
+			switch t := valueType.(type) {
+			case *btf.Struct:
+				// The values member pointing to an array of structs means we're expecting
+				// a map-in-map declaration.
+				if MapType(mapType) != ArrayOfMaps && MapType(mapType) != HashOfMaps {
+					return nil, errors.New("outer map needs to be an array or a hash of maps")
+				}
+				if inner {
+					return nil, fmt.Errorf("multi-level inner maps not supported")
+				}
+
+				// Name the inner map after the struct that declared it (can be anonymous).
+				// Pass the BTF spec from the parent object as parent and child need to be
+				// created from the same BTF blob to be valid (on kernels that support BTF).
+				innerMapSpec, err = mapSpecFromBTF(string(t.Name), t.Members, true, spec)
```

Comment:
I'd probably leave this unnamed. `MapSpec.InnerMap` is only ever used to create a temporary map that serves as a template to the kernel, so a name isn't very useful. There is also a subtlety here: is the name of the type really what the user would expect here?

Reply:
Yup, the type name is absolutely meaningless here. Thought I'd already removed it, thanks!